pkg/util/rbac/rbac.go

```diff
@@ -68,32 +68,90 @@ func ResourceRoleAssignmentWithName(roleID, spID, resourceType, resourceName, na
 // resource group to spID.  Argument spID must be a valid ARM expression, e.g.
 // "'foo'" or "concat('foo')".  Use this function in new ARM templates.
 func ResourceGroupRoleAssignment(roleID, spID string, condition ...interface{}) *arm.Resource {
-	return ResourceGroupRoleAssignmentWithName(roleID, spID, "guid(resourceGroup().id, '"+spID+"', '"+roleID+"')", condition)
+	return ResourceGroupRoleAssignmentWithName(roleID, spID, "guid(resourceGroup().id, '"+spID+"', '"+roleID+"')", condition...)
 }
 
-// ResourceGroupRoleAssignmentWithName returns a Resource granting roleID on the
-// current resource group to spID.  Arguments spID and name must be valid ARM
-// expressions, e.g. "'foo'" or "concat('foo')".  Use this function in ARM
-// templates which have already been deployed, to preserve the name and avoid a
-// RoleAssignmentExists error.
-func ResourceGroupRoleAssignmentWithName(roleID, spID string, name string, condition ...interface{}) *arm.Resource {
+func resourceGroupRoleAssignmentWithDetails(roleID, spID string, name string, dependsOn []string, subscriptionScope bool, condition ...interface{}) *arm.Resource {
+	var resourceIdFunction string
+	if subscriptionScope {
+		resourceIdFunction = "subscriptionResourceId"
+	} else {
+		resourceIdFunction = "resourceId"
+	}
 	r := &arm.Resource{
 		Resource: mgmtauthorization.RoleAssignment{
 			Name: to.StringPtr("[" + name + "]"),
 			Type: to.StringPtr("Microsoft.Authorization/roleAssignments"),
 			RoleAssignmentPropertiesWithScope: &mgmtauthorization.RoleAssignmentPropertiesWithScope{
 				Scope:            to.StringPtr("[resourceGroup().id]"),
-				RoleDefinitionID: to.StringPtr("[subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '" + roleID + "')]"),
+				RoleDefinitionID: to.StringPtr("[" + resourceIdFunction + "('Microsoft.Authorization/roleDefinitions', '" + roleID + "')]"),
 				PrincipalID:      to.StringPtr("[" + spID + "]"),
 				PrincipalType:    mgmtauthorization.ServicePrincipal,
 			},
 		},
 		APIVersion: azureclient.APIVersion("Microsoft.Authorization"),
 	}
-
+	if dependsOn != nil {
+		r.DependsOn = dependsOn
+	}
```

Comment:
If this required? If it is nil, it can be valid value aka empty?
```
r.DependsOn = dependsOn
```

Reply:
It could, and I believe it's valid. The reason I did it this way is it doesn't modify the existing deployment templates this way.

Reply:
I moved everything inside the Resource struct, so no separate r.DependsOn assignment is required.